File: dyn.ts
const mode = "safe";

type HasLength = { length: number };

type WithModeRequest<Result> = {
  mode: string;
  action: () => Result;
};

function reallyPerform<Task extends HasLength>(task: Task): number {
  console.log(`${mode}: ${task}`);
  return task.length;
}

function perform<Task extends HasLength>(task: Task): number {
  return reallyPerform(task);
}

function withMode<Result>(request: WithModeRequest<Result>): Result {
  const mode = request.mode;
  return request.action();
}

export function main() {
  console.log(
    perform("something"),
    withMode({ mode: "faster", action: () => perform("reliable") }),
    perform(["again"]),
  );
}

main();

File: dyn.scala
// https://scastie.scala-lang.org/nD6CaZNxRMiDfD59MD3fAw

case class Env(
  mode: String,
)

given Env = Env(mode = "safe")

def reallyPerform[A, Task <: Seq[A]](task: Task)(using env: Env) =
  println(s"${env.mode}: $task")
  task.length

def perform[A, Task <: Seq[A]](task: Task)(using env: Env) =
  reallyPerform(task)

// def withMode[T](mode: String)(action: (env: Env) ?=> T)(using env: Env) =
//   given Env = env.copy(mode = mode)
//   action

def withMode[Result](mode: String)(action: Env ?=> Result) =
  (env: Env) ?=> action(using env.copy(mode = mode))

@main def main() =
  var result = perform("something")
  result += withMode("faster") { perform("reliable") }
  result += perform(List("again"))
  println(s"result: $result")
end main
